src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,302 @@
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        async Task RemoveExpiredValuesPeriodically(TimeSpan interval)
+        {
+            while (true)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
+                await Task.Delay(interval).ConfigureAwait(false);
+            }
+        }
+
+        public void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        public bool SetValue(TKey key, TValue value, DateTime expirationTime)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (value == null)
+                throw LogHelper.LogArgumentNullException(nameof(value));
+
+            // item already expired
+            if (expirationTime < DateTime.UtcNow)
+                return false;
+
+            // just need to update value and move it to the top
+            if (_map.ContainsKey(key))
+            {
+                var newCacheItem = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
+                // make sure node hasn't been removed by a different thread
+                if (_map.TryGetValue(key, out var cacheItem))
+                {
+                    _eventQueue.Add(() =>
+                    {
+                        _doubleLinkedList.Remove(cacheItem);
+                        _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+
+                    });
+                }
+                else
+                {
+                    _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+                }
+
+                return true;
+            }
+            else
+            {
+                // if cache is full, then remove the least recently used node
+                if (_map.Count >= _capacity)
+                {                
+                    _eventQueue.Add(() =>
+                    {
+                        var lru = _doubleLinkedList.Last;
+                        _map.TryRemove(lru.Value.Key, out _);
+                        _doubleLinkedList.Remove(lru);
+                    });
+                }
+
+                // add a new node
+                var node = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
```

Comment:
>var node = new LRUCacheItem<TKey, TValue>(key, value, expirationTime); [](start = 16, length = 70)

seems to be the same item you are also creating above, perhaps move this out of the if/else statement since you are using it regardless?

Reply:
Good point; I've fixed this.https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1575/commits/38b55da469dcd67dfc96114d15fc2edf2a3802ba

I do believe the perf tests also report on memory, but I agree this is something we should look into if they don't. I believe Brent should have the results of the perf tests by tomorrow.

Do you mind elaborating on what you mean by "recreating the item in the scenario where it's not needed"? I'm not sure I follow.